envoy/buffer/buffer.h

```diff
@@ -458,6 +458,40 @@ class Instance {
     writeInt<ByteOrder::BigEndian, T, Size>(value);
   }
 
+  /**
+   * Copy multiple arguments to the buffer. The type of arguments must be automatically convertible
+   * (e.g. char*, char[], std::string) to `absl::string_view`.
+   * @param args A set of arguments with variable length and type. The arguments must be guaranteed
+   * to be safely convertible to `absl::string_view`.
+   * @return The total size of the data copied to the buffer.
+   */
+  template <bool ActionAfterAdd = true, class... Args> size_t addFragments(Args&&... args) {
+    size_t total_size_to_write = 0;
+
+    static_assert(sizeof...(args) >= 2, "At least two arguments");
+
+    constexpr auto write_helper = [](uint8_t** dst, absl::string_view v) {
+      memcpy(*dst, v.data(), v.size()); // NOLINT(safe-memcpy)
+      *dst += v.size();
+    };
+
+    total_size_to_write = (absl::string_view(args).size() + ...);
```

Comment:
I think that the multiple conversions of args to string_view on lines 478 and 482 may have some performance impact in the case where args contains "const char*".  I think that it may be possible to address this issue by changing the addFragments signature to:

template <bool ActionAfterAdd = true> size_t addFragments(absl::string_view&&... args) {

Reply:
Yeah that might make sense. It would be a little less syntactically convenient, but you don't need to use dynamic arrays. You can just use a begin/end at the interface, and write things like:
```
const absl::string_view args[] = {name, ": ", value, "\r\n"};
addFragments(args, args + arraysize(args));
```
this would also work for std::vector if that's more convenient, at the cost of some allocs.

Reply:
I can do a quick test for this. Because we need to consider the cost of building the container itself. 🤔 Or we can just ensure all args in the original `addFragments` must have `data()` and `size()` methods.

Reply:
I don't think you'll need to allocate any memory for the first option, a C Array like that will just extend the stack.

Reply:
Yes and I will run some test base the first option.

Reply:
For all `absl::string_view` args, the first option get almost same result with the current `addFragments` .

And if we change the method signature as `size_t addFragments(absl::Span<absl::string_view>)`, then we can just make it a normal virtual method. 🤔

Reply:
@jmarantz @antoniovicente I created a new implementation which take `absl::Span<absl::string_view>` as the parameter.

```
size_t addFragments(absl::Span<absl::string_view> fragments) {
  // ...
}
```

Any `std::vector<absl::string_view>`, `std::array<absl::string_view, N>` or `absl::string_view[]` can be used as the arguments. Even `std::initializer_list` can be accepted.

```

Buffer::OwnImpl buffer;

std::vector<absl::string_view> a = { "abc", "abc"};

std::array<absl::string_view, 2> b = {"abc", "efg"};

absl::string_view c[] = {"abc", "abc"};

buffer.addFragments(a);
buffer.addFragments(b);
buffer.addFragments({c, 2});
buffer.addFragments({"abc", "abc"});

```

Its performance is almost indistinguishable from previous template-based implementations, or even better. But we cannot use template parameters to disable watermark check. 

I think may be we can let `WatermarkBuffer` itself provide some methods in the new PR to temporarily disable watermark check.

Reply:
Cool! Span is the right construct for this; great choice!

WDYT of having the model for HTTP1 where you collect a single vector of Spans for all the fragments of all the headers, and use a single addSpans calls that does the watermark check on the aggregated size? Does that simplify things?

Reply:
> WDYT of having the model for HTTP1 where you collect a single vector of Spans for all the fragments of all the headers, and use a single addSpans calls that does the watermark check on the aggregated size? Does that simplify things?


Then we need create a vector to store all the key/values string view. Although frequent watermark checks will bring some overhead, I think copying all string view to the heap may bring more overhead. 

In addition, I haven't figured out how to deal with the situation where a large number of fragments are written at the same time in a single `addFragments`. If the back slice overflows, then we will downgrade to use `addImpl`.

Although our performance is definitely not worse than repeated `add()`, there is not much performance gain when overflow occurs.

Reply:
RE copying string-views; that's pretty small (16 bytes per).
RE heap: we have a limited number of headers, right? You could put a string_view[MAX_HEADERS] on the stack; no need to heap it.

RE add vs addFragments: as discussed in previous comments I'm suggesting we not have the policy of bailing into *add* mode when you spill out of the back slice. Instead, with the giant array of fragments, you can figure out before adding anything exactly how many bytes are needed, set up the slices as required, and do the watermarking once. I think it will be a cleaner/faster model.

Reply:
It's not just an optimization tweak -- it's a much simpler model, isn't it?

With the API that you have you have to have the caller do addFragments calls, and then end with an add(). That's potentially error-prone and I believe @ggreenway has commented about this.

Moreover the current API you have, as @antoniovicente observed, may have excess conversions from const char* to string_view which would result in extra calls to strlen.

With the Span API you can make the interface much simpler *and* more performant. So although I know this PR has gone on for a while, we iterated quite a bit and learned a lot and got some really good data. We are trying to make Envoy fast *and* safe, and it's very useful to do this well.

Reply:
@jmarantz I am a little confused. I have updated this PR to use new `absl::Span<absl::string_view>` API. Because as you said, it's cleaner, simper, and even faster. So the problems of @ggreenway and @antoniovicente have been solved. 🤣 Do you noticed the latest commit?

The optimization I mentioned above is mean the special case where the back slice has no continuous memory. We still can do some more perf improvements in that case.